src/librustc_allocator/expand.rs

```diff
@@ -142,23 +144,18 @@ impl<'a> Folder for ExpandAllocatorDirectives<'a> {
         let module = f.cx.monotonic_expander().fold_item(module).pop().unwrap();
 
         // Return the item and new submodule
-        smallvec![item, module]
+        Some(vec![item.clone(), module])
```

Comment:
It would be nice to find a way to "steal" the item when folding `item -> [item, extra_items]` rather than clone it, this should be a common enough scenario.

Looks like other cases are covered well.
`item -> item` - trivial
`item -> []` - trivial
`item -> [totally, unrelated, items]` - you can't do much with this.

Reply:
Alternatively, `one -> many` folders can return 3-variant enum instead of 2-variant `Option`.
```rust
enum Action {
    Remove, // item -> []
    Add(Vec<Item>), // item -> [item, extra_items], extra_items can be empty
    Replace(Vec<Item>), // item -> [items]
}
```

Reply:
(Or `Add(vec![])` can be turned into an extra variant `Reuse` for clarity.)

Reply:
Yeah, I think enum would be better for both optimization and convenience of writing specific folders.